### PR TITLE
MEL supporting handling of rasterizing polygons with holes

### PR DESCRIPTION
**Pull request by marinhekman**
4 main changes:
* MEL supporting handling of rasterizing polygons with holes
* one-time "isonline" polling towards the server. now we are just checking http status code 401, throwing an ApiUnauthorizedWebException
* limited polling towards the server to retrieve the "current month". the server will push the next month towards the watchdog, which is forwarded to the simulations using the pipes
* Created a ConsoleTextWriter.cs that is injected into Console to be able to format the output messages to console. E.g. each simulation output it prefixed, e.g. MEL:

changes:
- added interface IUpdateMonthReceiver. extended IAPiConnnector to implement it as well as ITokenReceiver
- Added a ApiConnectorBase used by the simulation to create their api connector having common functionality
- Renamed APITokenHandler to CommunicationPipeHandler since it handle the update of the month now as well through the pipes
- added missing ? for variables being handled as nullable
- fixed output padding using PadRight
- added SmartFormat lib
- added pdb file handling in build.bat

### CODE REVIEW

File: MSWSupport/MSWSupport/CommunicationPipeHandler.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+				m_updateMonthReceiver?.UpdateMonth(m_currentMonth);
+			} while (!reader.EndOfStream);
+		}
+
+		public void Dispose()
+		{
+			m_communicationPipe.Dispose();
+		}
+	}
+}
```
Comment (hwarmelink) [inline MSWSupport/MSWSupport/CommunicationPipeHandler.cs]:
```suggestion
}

```

Comment (hwarmelink) [reply MSWSupport/MSWSupport/CommunicationPipeHandler.cs]: ha! I finally get to do this to you too! ;) hihihi (only joking)

File: MSW/MSW/RestEndpointSetMonth.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			Encoding encoding = a_response.ContentEncoding ?? Encoding.UTF8;
+			byte[] buffer = encoding.GetBytes(responseString);
+			a_response.ContentLength64 = buffer.Length;
+			Stream output = a_response.OutputStream;
+			output.Write(buffer, 0, buffer.Length);
+
+			output.Close();
+		}
+	}
+}
```
Comment (hwarmelink) [inline MSW/MSW/RestEndpointSetMonth.cs]:
```suggestion
}

```

File: MEL/MEL/ApiMspServer.cs [hunk 3]
```diff
@@ -144,78 +105,42 @@ public void NotifyTickDone()
...
-		{
-			return APIRequest.Perform(m_ServerUrl, apiUrl, m_CurrentAccessToken, postValues, out result);
-		}
-
-		public void UpdateAccessToken(string newAccessToken)
-		{
-			m_CurrentAccessToken = newAccessToken;
-		}
 	}
 }
```
Comment (hwarmelink) [inline MEL/MEL/ApiMspServer.cs]:
```suggestion
}

```

File: REL/REL/API/MSPAPIDebug.cs [hunk 4]
```diff
@@ -37,5 +37,10 @@ public void UpdateAccessToken(string newAccessToken)
 		{
 			throw new System.NotImplementedException();
 		}
+
+		public void UpdateMonth(int month)
+		{
+			throw new System.NotImplementedException();
+		}
 	}
 }
```
Comment (hwarmelink) [inline REL/REL/API/MSPAPIDebug.cs]:
```suggestion
}

```

File: MSWSupport/MSWSupport/ITokenReciever.cs [hunk 5]
```diff
@@ -0,0 +1,6 @@
+﻿namespace MSWSupport;
+
+public interface ITokenReceiver
+{
+	void UpdateAccessToken(string newAccessToken);
+}
```
Comment (hwarmelink) [inline MSWSupport/MSWSupport/ITokenReciever.cs]:
```suggestion
}

```

**Comment by hwarmelink** [comment]
I believe you can specify an 'api_version' argument in the build.bat script to ensure appropriate version.txt files get created as well (I'm referring here to my comment to bump the simulation versions in the linked MSPChallenge-Server PR). I suggest you set version 1.1.0 for these updates.

**Comment by hwarmelink** [comment]
Nice work with the clean-up of existing code you've squeezed into this work as well.

**Review by hwarmelink — APPROVED** [review]
Check my comments please before you merge though.

File: MEL/MEL/APILayerGeometryData.cs [hunk 6]
```diff
@@ -5,6 +5,6 @@ namespace MEL
 	public class APILayerGeometryData
 	{
 		public string geotype { get; set; }
-		public List<List<double[]>> geometry { get; set; }
+		public List<List<List<double[]>>> geometry { get; set; }
 	}
 }
```
Comment (marinhekman) [inline MEL/MEL/APILayerGeometryData.cs]:
```suggestion
}

```

**Pull request merged**